src/vmm/src/device_manager/acpi.rs

```diff
@@ -1,21 +1,28 @@
 // Copyright 2024 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 // SPDX-License-Identifier: Apache-2.0
 
+use std::sync::{Arc, Mutex};
+
 use acpi_tables::{aml, Aml};
 use kvm_ioctls::VmFd;
 
+use crate::devices::acpi::cpu_container::CpuContainer;
 use crate::devices::acpi::vmgenid::VmGenId;
 
 #[derive(Debug)]
 pub struct ACPIDeviceManager {
     /// VMGenID device
     pub vmgenid: Option<VmGenId>,
+    pub cpu_container: Option<Arc<Mutex<CpuContainer>>>,
```

Comment:
mh, will we ever _not_ have a CpuContainer?

Reply:
Hmm. Well in the case that you don't have ACPI enabled, I suppose not? But you also wouldn't have the whole ACPIDevMgr, so that's a moot point I suppose. I'd need to look at how snapshot restore works with this, I imagine it's a fairly simple change, but I need to have a look

Reply:
Do we support disabling ACPI (I genuinely don't remember)?

Reply:
Well for the time being we have the no-acpi config for 5.10

Reply:
But does using that actually cause Firecracker to disable ACPI features, or does it mean Firecracker writes the ACPI tables anyway, but the guest simply ignores them?

Reply:
I wasn't sure about this to be honest, so I just tested it. It looks like the latter is right. After running the API call there are no new CPUs present to the guest. However, it does lead to an interesting edge case: the vCPU threads are made on the VMM side, but the CPUs are never usable. I'm not sure the best way to handle this edge case, I can potentially look and see if there is a way to query the kernel about ACPI support? Then an error can be returned.

Reply:
> Would it maybe be better just to assume there is ACPI support, and ignore cases where there isn't?

Yeah, I think so. Booting without ACPI is deprecated already, so in the (near) future there won't even be the possibility of not having ACPI support anymore.

Reply:
Okay cool. In that case, I don't see any reason why the CpuContainer shouldn't exist. I'll modify it as such.

Reply:
Confirming that Firecracker always writes ACPI tables in guest memory. The `no-acpi` kernel is a guest kernel that does not support ACPI. In these cases the guest just boots with the "legacy" way of MPTABLE and command line arguments. As Patrick correctly said, we will eventually drop support for these, i.e. we will not write the MPTABLE in guest memory, neither pass VirtIO devices via the command line.

Reply:
I'm thinking of undoing this change, it's causing issues that aren't immediately obvious. When `CpuContainer` is not wrapped in an option, it must be constructed and passed as an argument into the constructor of the `ACPIDeviceManager`. The issue with this is that `ACPIDeviceManager` must be constructed in the `create_vmm_and_vcpus` method, this means that `CpuContainer::new` must also occur in this method. The issue with this is that `CpuContainer::new` allocates MMIO memory and a GSI for ACPI. 

To make this possible `setup_interrupt_controller` must be moved inside `create_vmm_and_vcpus` (so GSIs can be allocated), and the `BootTimer` also has to be moved inside the same method, as it has to occupy the first MMIO address in order to work properly. This has the potential to cause confusion, as now devices are being registered in a place separate from the rest of them, and it leads to `create_vmm_and_vcpus` having unexpected side-effects. By making `CpuContainer` optional again, both issues can be sorted. The `BootTimer` device can be registered first, and `setup_interrupt_controller` can be moved back to it's original location, in `build_microvm_for_boot`.